core/util/helpers.php

```diff
@@ -16,6 +31,109 @@ function contains($haystack, $needle) {
+                for ($i = 0; $i < count($routeArr); $i++) {
+                    if ($routeArr[$i] != '*' && $routeArr[$i] != '**' &&
+                        $routeArr[$i] != $urlArr[$i]) {
+                        $match = false;
+                        break;
+                    }
+                }
+                # Set new permission if there was a match
+                if ($match) {
+                    $newPerm = $perm;
+                    break;
+                }
+            } else if (count($routeArr) < count($urlArr)) {
+                $match = true;
+                # Match each part of the URL with the permission route
+                for ($i = 0; $i < count($urlArr); $i++) {
+
+                    # Route not long enough
+                    if (!isset($routeArr[$i])) {
+                        $match = false;
+                        break;
+                    }
+
+                    # Matches with everything after this point, break
+                    if ($routeArr[$i] == '**') {
+                        break;
+                    }
+
+                    # Doesn't match
+                    if ($routeArr[$i] != $urlArr[$i]) {
+                        $match = false;
+                        break;
+                    }
+                }
+                # Set new permission if there was a match
+                if ($match) {
+                    $newPerm = $perm;
+                    break;
+                }
+            }
+        }
+        if (empty($newPerm)) {
+            echo 'WARNING: No page permission defined for ' . $url;
+        } else {
+            $per = $newPerm;
+        }
+    }
+
+    # once a page permission is found, verify authorization
+    $opts = explode(' ', $per);
+    $type = $opts[0]; # either ALLOW or DENY
+    array_shift($opts);
+    $roles = $opts; # roles to verify
+    if ($type == 'ALLOW') {
+        if ($roles[0] == '*') {             # allows all connections
+            return;
+        } else if ($roles[0] == 'auth') {   # allows authorized connections
+            authorizedPage();
+        } else {                            # allows the listed roles
+            $allowRoles = 0;
+            foreach ($roles as $r) {
+                $allowRoles |= Role::roleFromPermissionLevel($r);
+            }
+            requireRole($allowRoles);
+        }
+    } else if ($type == 'DENY') {           # denies the listed roles
+        if ($roles[0] == '*') {
+            notAuthorized();
+        } else {
+            $denyRoles = 0;
+            foreach ($roles as $r) {
+                $denyRoles |= Role::roleFromPermissionLevel($r);
+            }
+            preventRole($denyRoles);
+        }
+    } else {
+        throw new InvalidArgumentException("Neither 'ALLOW' nor 'DENY' was specified in permission for '$url'");
+    }
+}
+
+/**
  * Call this function atop each page that requires
  * authorization
  */
```

Comment:
Is this function still needed?

Reply:
Yeah it's used on `line 119` of this file